server/lib/actors/hot_test.go

```diff
@@ -0,0 +1,250 @@
+	switch btyp {
+	case proto.TopicListRequest_Rising:
+		return testInmemVoteFetcher()
+	case proto.TopicListRequest_Hot, proto.TopicListRequest_Flame:
+		return testInmemVoteParentFetcher(start, end, btyp)
+	default:
+		return nil, fmt.Errorf("invalid fetchType: %v", btyp)
+	}
+}
+
+func testInmemPersister(at time.Time, s *hotbucketStore) error {
+	return nil
+}
+
+func estimateResults(btyp proto.TopicListRequest_BucketType, typ proto.TopicListRequest_QueryType, start, count uint64) []HotEntry {
+	var begin, end time.Time;
+	end = getClock()
+	switch typ {
+	case proto.TopicListRequest_Hour:
+		begin = end.Add(-1 * time.Hour)
+	case proto.TopicListRequest_Day:
+		begin = end.Add(-24 * time.Hour)
+	case proto.TopicListRequest_Week:
+		begin = end.Add(-7 * 24 * time.Hour)
+	case proto.TopicListRequest_AllTime:
+		begin = time.Time{}
+	}
+	tmp := NewHotBucket(ENTRY_PER_FETCH)
+	for _, e := range estimates {
+		switch btyp {
+		case proto.TopicListRequest_Rising:
+			//rising is calculated for votes which is done in specified term. so add all if its in specified term.
+			if e.fetchAt.UnixNano() > begin.UnixNano() && e.fetchAt.UnixNano() <= end.UnixNano() {
+				tmp.addResults(e.results)
+			}
+		case proto.TopicListRequest_Hot, proto.TopicListRequest_Flame:
+			//Hot and Flame is calculated for objects which created in specified term, so exclude objects which is out of range
+			tmp.addResultsWithFilter(e.results, func (ent *HotEntry) bool {
+				//log.Printf("addResultsWithFilter: %v %v %v %v %v %v", i, ent, begin, end, ent.within(begin, end), yue.DateByUUID(yuep.UUID(ent.id)))
+				return ent.within(begin, end)
+			})
+		}
+		//week/alltime: emurate filter on making 12h summery
+		if e.fetchAt.UnixNano() == lastFilter.UnixNano() && 
+			(typ == proto.TopicListRequest_Week || typ == proto.TopicListRequest_AllTime) {
+			log.Printf("emurate filter on summery %v %v %v", lastFilter, getClock(), tmp.total.entries)
+			tmp = filterSummery(btyp, tmp)
+			log.Printf("emurate filter on summery done %v", tmp.total.entries)
+		}
+	}
+
+	switch btyp {
+	case proto.TopicListRequest_Rising, proto.TopicListRequest_Hot:
+		//log.Printf("sort tmp: %v", tmp.total.entries)
+		tmp = tmp.sort(btyp).truncate(ENTRY_PER_FETCH)
+	case proto.TopicListRequest_Flame:
+		tmp = tmp.filter(func (e *HotEntry) bool {
+			return e.flamy()
+		}).sort(btyp).truncate(ENTRY_PER_FETCH)
+		//log.Printf("sort tmp: %v", tmp.total.entries)
+	}
+	last := int(start+count)
+	if last >= len(tmp.total.entries) {
+		last = len(tmp.total.entries)
+	}
+	//log.Printf("estimateResults %v %v %v %v", tmp.total.entries, start, count, last)
+	return tmp.total.entries[start:last]
+
+}
+
+func NewTestHotActor(locale string, updateSpan time.Duration, blen []int) (*HotActor, error) {
+	return NewHotActorWithDetail(locale, updateSpan, blen, true, testInmemFetcher, testInmemPersister);
+}
+
+func TestHotActor(t *testing.T) {
```

Comment:
__[gometalinter]__
* [gocyclo] cyclomatic complexity 15 of function TestHotActor() is high (> 10)